File: templates/express-js/src/api/schemas/pagination.ts
import { z } from "zod";

// Default values for pagination
export const DEFAULT_PAGE = 1;
export const DEFAULT_PAGE_SIZE = 100;

/** Query parameters for paginated routes */
export const PaginatedQueryParamsSchema = z.object({
  /** The page to return (minimum value: 1) */
  page: z.coerce.number().int().min(1).optional().default(DEFAULT_PAGE),
  /** The number of items to return per page (minimum value: 1) */
  pageSize: z.coerce.number().int().min(1).optional().default(DEFAULT_PAGE_SIZE),
});

/** Body parameters for paginated routes */
export const PaginatedBodyParamsSchema = z.object({
  /** The page to return (minimum value: 1) */
  page: z.number().int().min(1).optional().default(DEFAULT_PAGE),
  /** The number of items to return per page (minimum value: 1) */
  pageSize: z.number().int().min(1).optional().default(DEFAULT_PAGE_SIZE),
});

/** Details about the paginated results */
export const PaginationInfoSchema = z.object({
  /** Current page number (indexing starts at 1) */
  page: z.number().int().min(1),
  /** Number of items per page */
  pageSize: z.number().int().min(1),
  /** Total number of items across all pages */
  totalItems: z.number().int().optional(),
  /** Total number of pages */
  totalPages: z.number().int().optional(),
});

// Export types inferred from the schemas
export type PaginatedQueryParams = z.infer<typeof PaginatedQueryParamsSchema>;
export type PaginatedBodyParams = z.infer<typeof PaginatedBodyParamsSchema>;
export type PaginationInfo = z.infer<typeof PaginationInfoSchema>;

File: lib/python-sdk/common_grants_sdk/schemas/pagination.py
"""Paginated models for the CommonGrants API."""

from typing import Generic, TypeVar

from pydantic import BaseModel, Field

T = TypeVar("T")


class PaginatedBase(BaseModel):
    """Parameters for pagination."""

    page: int = Field(
        default=1,
        description="The page number to retrieve",
        ge=1,
    )
    page_size: int = Field(
        default=10,
        alias="pageSize",
        description="The number of items per page",
        ge=1,
    )

    model_config = {"populate_by_name": True}


class PaginatedBodyParams(PaginatedBase):
    """Parameters for pagination in the body of a request."""


class PaginatedQueryParams(PaginatedBase):
    """Parameters for pagination in a request query."""


class PaginatedResultsInfo(PaginatedBase):
    """Information about the pagination of a list."""

    total_items: int = Field(
        ...,
        alias="totalItems",
        description="The total number of items",
    )
    total_pages: int = Field(
        ...,
        alias="totalPages",
        description="The total number of pages",
    )

    model_config = {"populate_by_name": True}


class PaginatedItems(BaseModel, Generic[T]):
    """A paginated list of items."""

    items: list[T] = Field(..., description="The list of items")
    paginated_info: PaginatedResultsInfo = Field(
        ...,
        alias="paginatedInfo",
        description="Information about the pagination",
    )

    model_config = {"populate_by_name": True}
